docs/source/inspect.rst

```diff
@@ -0,0 +1,150 @@
+Inspecting Types
+----------------
+
+.. currentmodule:: msgspec.inspect
+
+.. warning::
+
+   This module is experimental. While we don't expect any breaking changes, we
+   also don't promise not to break things between releases while this interface
+   stabilizes.
+
+``msgspec`` provides type-introspection support, which can be used to build
+tooling on top of msgspec-compatible types. Possible use cases include:
+
+- Generating OpenAPI_ specifications from msgspec-compatible types (note that
+  the builtin :doc:`jsonschema` support may be a better starting point for
+  this).
+- Generating example instances of types for testing or documentation purposes
+- Integration with hypothesis_ for testing
+
+The main function here is `msgspec.inspect.type_info` for converting a type
+annotation into a corresponding `msgspec.inspect.Type` object. There's also
+`msgspec.inspect.multi_type_info` which converts an iterable of annotations;
+this function is more efficient than calling `type_info` in a loop.
+
+.. code-block:: python
+
+    >>> import msgspec
+
+    >>> msgspec.inspect.type_info(bool)
+    BoolType()
+
+    >>> msgspec.inspect.type_info(int)
+    IntType(gt=None, ge=None, lt=None, le=None, multiple_of=None)
+
+    >>> msgspec.inspect.type_info(list[int])  # nested types are traversed
+    ListType(
+        item_type=IntType(gt=None, ge=None, lt=None, le=None, multiple_of=None),
+        min_length=None,
+        max_length=None
+    )
+
+    >>> msgspec.inspect.multi_type_info([bool, int])  # inspect multiple types
+    (BoolType(), IntType(gt=None, ge=None, lt=None, le=None, multiple_of=None))
+
+
+Types with :doc:`constraints` will include the constraint information as well:
+
+.. code-block:: python
+
+    >>> from typing import Annotated
+
+    >>> from msgspec import Meta
+
+    >>> PositiveInt = Annotated[int, Meta(gt=0)]
+
+    >>> msgspec.inspect.type_info(PositiveInt)
+    IntType(gt=0, ge=None, lt=None, le=None, multiple_of=None)
+
+Compound types like :doc:`structs` are also supported:
+
+.. code-block:: python
+
+    >>> class User(msgspec.Struct):
+    ...     name: str
+    ...     groups: list[str] = []
+    ...     email: str | None = None
+
+    >>> msgspec.inspect.type_info(User)
```

Comment:
What would be the output of complex types? for example:

```python
dict[str, list[dict[str, tuple[str, str]]]]
deque[str | float | tuple[str, str], list[dict[str,list[int]]]
```

?

Additionally, it would be helpful to know if a field is `allow_none` without inspecting the Union to see if its a union of None.

Reply:
> What would be the output of complex types?

```python
In [1]: import msgspec, rich

In [2]: t = msgspec.inspect.type_info(dict[str, list[dict[str, tuple[str, str]]]])

In [3]: rich.print(t)
DictType(
    key_type=StrType(min_length=None, max_length=None, pattern=None),
    value_type=ListType(
        item_type=DictType(
            key_type=StrType(min_length=None, max_length=None, pattern=None),
            value_type=TupleType(item_types=(StrType(min_length=None, max_length=None, pattern=None), StrType(min_length=None, max_length=None, pattern=None))),
            min_length=None,
            max_length=None
        ),
        min_length=None,
        max_length=None
    ),
    min_length=None,
    max_length=None
)
```

> Additionally, it would be helpful to know if a field is allow_none without inspecting the Union to see if its a union of None.

I've added an `include_none` property to `UnionType` nodes to handle this.